singularity-eos/eos/eos_eospac.hpp

```diff
@@ -72,103 +72,1156 @@ class EOSPAC : public EosBase<EOSPAC> {
-  //                                           ConstRealIndexer &&temperatures,
-  //                                           RealIndexer &&gm1s,
-  //                                           const int num,
-  //                                           LambdaIndexer &&lambdas) const;
-
-  // template<typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
-  // inline
-  // void GruneisenParamFromDensityInternalEnergy(ConstRealIndexer &&rhos,
-  //                                              ConstRealIndexer &&sies,
-  //                                              RealIndexer &&gm1s,
-  //                                              const int num,
-  //                                              LambdaIndexer &&lambdas) const;
-
-  // template<typename RealIndexer, typename LambdaIndexer>
-  // inline
-  // void FillEos(RealIndexer &&rhos, RealIndexer &&temps, RealIndexer &&energies,
-  //              RealIndexer &&presses, RealIndexer &&cvs, RealIndexer &&bmods,
-  //              const int num, const unsigned long output,
-  //              LambdaIndexer &&lambdas) const;
+  // EOSPAC vector implementations
+  template <int CHUNK_SIZE = 64, typename RealIndexer, typename ConstRealIndexer,
+            typename LambdaIndexer>
+  inline void
+  TemperatureFromDensityInternalEnergy(ConstRealIndexer &&rhos, ConstRealIndexer &&sies,
+                                       RealIndexer &&temperatures, const int num,
+                                       LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    int nchunks = (num / CHUNK_SIZE) + (num % CHUNK_SIZE ? 1 : 0);
+    std::array<EOS_REAL, CHUNK_SIZE> R, E, T, dTdr, dTde;
+
+    for (int ichunk = 0; ichunk < nchunks; ++ichunk) {
+      int istart = ichunk * CHUNK_SIZE;
+      int iend = std::min(num, istart + CHUNK_SIZE);
+      int nelements = iend - istart;
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        R[i] = rhos[j];
+        E[i] = sieToSesame(sies[j]);
+      }
+
+      EOS_INTEGER table = TofRE_table_;
+      eosSafeInterpolate(&table, nelements, R.data(), E.data(), T.data(), dTdr.data(),
+                         dTde.data(), "TofRE", Verbosity::Quiet);
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        temperatures[j] = temperatureFromSesame(T[i]);
+      }
+    }
+  }
+
+  template <typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+  inline void
+  TemperatureFromDensityInternalEnergy(ConstRealIndexer &&rhos, ConstRealIndexer &&sies,
+                                       RealIndexer &&temperatures, Real *scratch,
+                                       const int num, LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    EOS_REAL *R = scratch;
+    EOS_REAL *E = scratch + num;
+    EOS_REAL *T = scratch + 2 * num;
+    EOS_REAL *dTdr = scratch + 3 * num;
+    EOS_REAL *dTde = scratch + 4 * num;
```

Comment:
We should probably have a way to compute the required scratch space and hand that to the user. I'm not sure the best way to do this though.

Reply:
Actually, can't we just use the `lambdas` variable for scratch? EOSPAC currently says it doesn't need any lambdas, but we could set that requirement to 8 to account for the largest possible scratch usage.

In fact, we could also overload `nlambdas()` to take an integer corresponding to the number of points to ease the calculation of lambda arrays for spiner as well.

Reply:
The required scratch space is `nscratch() * elements * sizeof(Real)`. We could have a helper function that just takes `nelements` and returns the number of bytes.

Reply:
One possible solution for returning the exact scratch space requirement per call is having a function that returns the right value for each member function pointer.

Reply:
@Yurlungur Added a `max_scratch_size(nelements)` method